Content.Shared/Decals/SharedDecalSystem.cs

```diff
@@ -0,0 +1,153 @@
+using System;
+using System.Collections.Generic;
+using System.Diagnostics.CodeAnalysis;
+using Robust.Shared;
+using Robust.Shared.Configuration;
+using Robust.Shared.GameObjects;
+using Robust.Shared.IoC;
+using Robust.Shared.Map;
+using Robust.Shared.Maths;
+using Robust.Shared.Prototypes;
+
+namespace Content.Shared.Decals
+{
+    public abstract class SharedDecalSystem : EntitySystem
+    {
+        [Dependency] protected readonly IPrototypeManager PrototypeManager = default!;
+        [Dependency] private readonly IConfigurationManager _configurationManager = default!;
+        [Dependency] protected readonly IMapManager MapManager = default!;
+
+        protected readonly Dictionary<GridId, Dictionary<uint, Vector2i>> ChunkIndex = new();
+
+        public const int ChunkSize = 32;
+        public static Vector2i GetChunkIndices(Vector2 coordinates) => new ((int) Math.Floor(coordinates.X / ChunkSize), (int) Math.Floor(coordinates.Y / ChunkSize));
+
+        private float _viewSize;
+
+        public override void Initialize()
+        {
+            base.Initialize();
+
+            SubscribeLocalEvent<GridInitializeEvent>(OnGridInitialize);
+            _configurationManager.OnValueChanged(CVars.NetMaxUpdateRange, OnPvsRangeChanged, true);
+        }
+
+        public override void Shutdown()
+        {
+            base.Shutdown();
+            _configurationManager.UnsubValueChanged(CVars.NetMaxUpdateRange, OnPvsRangeChanged);
+        }
+
+        private void OnPvsRangeChanged(float obj)
+        {
+            _viewSize = obj * 2f;
+        }
+
+        private void OnGridInitialize(GridInitializeEvent msg)
+        {
+            var comp = EntityManager.EnsureComponent<DecalGridComponent>(MapManager.GetGrid(msg.GridId).GridEntityId);
+            ChunkIndex[msg.GridId] = new();
```

Comment:
The mix of dictionary and component activates my almonds.

Reply:
sir its called an index. that being said this one doesnt get rebuilt after loading so i should probably do that